gobblin-modules/gobblin-helix/src/main/java/gobblin/metastore/ZkStateStore.java

```diff
@@ -0,0 +1,279 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.metastore;
+
+import com.google.common.base.Predicate;
+import com.google.common.base.Predicates;
+import com.google.common.base.Strings;
+import com.google.common.collect.Lists;
+import gobblin.configuration.State;
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.DataInputStream;
+import java.io.DataOutput;
+import java.io.DataOutputStream;
+import java.io.EOFException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.zip.GZIPInputStream;
+import java.util.zip.GZIPOutputStream;
+import org.I0Itec.zkclient.serialize.ZkSerializer;
+import org.apache.commons.lang.ArrayUtils;
+import org.apache.hadoop.io.Text;
+import org.apache.helix.AccessOption;
+import org.apache.helix.manager.zk.ByteArraySerializer;
+import org.apache.helix.store.HelixPropertyStore;
+import org.apache.helix.store.zk.ZkHelixPropertyStore;
+
+
+/**
+ * An implementation of {@link StateStore} backed by ZooKeeper.
+ *
+ * <p>
+ *
+ *     This implementation stores serialized {@link State}s as a blob in ZooKeeper in the Sequence file format.
+ *     The ZK path is in the format /STORE_ROOT_DIR/STORE_NAME/TABLE_NAME.
+ *     State keys are state IDs (see {@link State#getId()}), and values are objects of {@link State} or
+ *     any of its extensions. Keys will be empty strings if state IDs are not set
+ *     (i.e., {@link State#getId()} returns <em>null</em>). In this case, the
+ *     {@link ZkStateStore#get(String, String, String)} method may not work.
+ * </p>
+ *
+ * @param <T> state object type
+ **/
+public class ZkStateStore<T extends State> implements StateStore<T> {
+  // Class of the state objects to be put into the store
+  private final Class<T> stateClass;
+  private final HelixPropertyStore<byte[]> propStore;
+  private final boolean compressedValues;
+
+  /**
+   * State store that stores instances of {@link State}s in a ZooKeeper-backed {@link HelixPropertyStore}
+   * @param connectString ZooKeeper connect string
+   * @param storeRootDir The root directory for the state store
+   * @param compressedValues should values be compressed for storage?
+   * @param stateClass The type of state being stored
+   * @throws IOException
+   */
+  public ZkStateStore(String connectString, String storeRootDir, boolean compressedValues, Class<T> stateClass)
+      throws IOException {
+    this.compressedValues = compressedValues;
+    this.stateClass = stateClass;
+
+    ZkSerializer serializer = new ByteArraySerializer();
+    propStore = new ZkHelixPropertyStore<byte[]>(connectString, serializer, storeRootDir);
```

Comment:
There is no additional configuration of the timeouts allowed with the Helix property store.  It creates the ZkClient with session timeout of 30 seconds and connect timeout of 60 seconds.
    this._zkclient = new ZkClient(zkAddress, 30000, '\uea60', serializer);
    this._zkclient.waitUntilConnected(60000L, TimeUnit.MILLISECONDS);